src/Sparrow/Json/ArenaMemoryAllocator.cs

```diff
@@ -0,0 +1,124 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using NLog;
+
+namespace Sparrow.Json
+{
+    public unsafe class ArenaMemoryAllocator : IDisposable
+    {
+        private byte* _ptrStart;
+        private byte* _ptrCurrent;
+
+        private int _allocated;
+        private int _used;
+
+        private List<IntPtr> _unusedBuffers;
+
+        private bool _isDisposed;
+        private static readonly Logger _log = LogManager.GetLogger(nameof(ArenaMemoryAllocator));
+        
+        public ArenaMemoryAllocator(int initialSize = 1024 * 1024)
+        {
+            _ptrStart = _ptrCurrent = (byte*)Marshal.AllocHGlobal(initialSize).ToPointer();
+            _allocated = initialSize;
+            _used = 0;
+
+            _log.Debug($"ArenaMemoryAllocator was created with initial capacity of {initialSize:#,#;;0} bytes");
+        }
+
+        public AllocatedMemoryData Allocate(int size)
+        {
+            if (_isDisposed)
+                ThrowAlreadyDisposedException();
+
+            if (_used + size > _allocated)
+                GrowArena(size);
+
+            var allocation = new AllocatedMemoryData()
+            {
+                SizeInBytes = size,
+                Address = new IntPtr(_ptrCurrent)
+            };
+
+            _ptrCurrent += size;
+            _used += size;
+
+            _log.Debug($"ArenaMemoryAllocator allocated {size:#,#;;0} bytes");
```

Comment:
This will cause allocations. Always use checks before constructing strings for logs. The rest seems OK. I am operating under the assumption this is NOT a high traffic operation.